sdks/java/extensions/sql/src/main/java/org/apache/beam/sdk/extensions/sql/meta/provider/datastore/DataStoreV1Table.java

```diff
@@ -0,0 +1,479 @@
+
+    this.projectId = matcher.group("projectId");
+    this.kind = matcher.group("kind");
+  }
+
+  @Override
+  public PCollection<Row> buildIOReader(PBegin begin) {
+    Query.Builder q = Query.newBuilder();
+    q.addKindBuilder().setName(kind);
+    Query query = q.build();
+
+    DatastoreV1.Read readInstance =
+        DatastoreIO.v1().read().withProjectId(projectId).withQuery(query);
+
+    PCollection<Entity> readEntities = readInstance.expand(begin);
+
+    return readEntities.apply(EntityToRow.create(getSchema(), keyField));
+  }
+
+  @Override
+  public POutput buildIOWriter(PCollection<Row> input) {
+    return input
+        .apply(RowToEntity.create(keyField, kind))
+        .apply(DatastoreIO.v1().write().withProjectId(projectId));
+  }
+
+  @Override
+  public IsBounded isBounded() {
+    return IsBounded.BOUNDED;
+  }
+
+  @Override
+  public BeamTableStatistics getTableStatistics(PipelineOptions options) {
+    long count =
+        DatastoreIO.v1().read().withProjectId(projectId).getNumEntities(options, kind, null);
+
+    if (count < 0) {
+      return BeamTableStatistics.BOUNDED_UNKNOWN;
+    }
+
+    return BeamTableStatistics.createBoundedTableStatistics((double) count);
+  }
+
+  /**
+   * A {@code PTransform} to perform a conversion of {@code PCollection<Entity>} to {@code
+   * PCollection<Row>}.
+   */
+  public static class EntityToRow extends PTransform<PCollection<Entity>, PCollection<Row>> {
+    private final Schema schema;
+    private final String keyField;
+
+    private EntityToRow(Schema schema, String keyField) {
+      this.schema = schema;
+      this.keyField = keyField;
+
+      if (schema.getFieldNames().contains(keyField)
+          && !schema.getField(keyField).getType().getTypeName().equals(TypeName.BYTES)) {
+        throw new IllegalStateException(
+            "Field `"
+                + keyField
+                + "` should of type `VARBINARY`. Please change the type or specify a field to store the KEY value in via TableProperties.");
+      }
+    }
+
+    /**
+     * Create a PTransform instance.
+     *
+     * @param schema {@code Schema} of the target row.
+     * @return {@code PTransform} instance for Entity to Row conversion.
+     */
+    public static EntityToRow create(Schema schema) {
+      LOGGER.info(
+          "VARBINARY field to store KEY was not specified, using default value: `"
+              + DEFAULT_KEY_FIELD
+              + "`.");
+      return new EntityToRow(schema, DEFAULT_KEY_FIELD);
+    }
+
+    /**
+     * Create a PTransform instance.
+     *
+     * @param schema {@code Schema} of the target row.
+     * @param keyField A name of the row field to store the {@code Key} in.
+     * @return {@code PTransform} instance for Entity to Row conversion.
+     */
+    public static EntityToRow create(Schema schema, String keyField) {
+      LOGGER.info("VARBINARY field to store KEY was specified, using value: `" + keyField + "`.");
```

Comment:
This is probably too verbose and possibly confusing, as this will be printed even if the field wasn't specified.

Reply:
Simplified and moved this `info` message.
Should be displayed when `keyField` is one of the schema fields.